coordinator/src/main/scala/filodb.coordinator/IngestionActor.scala

```diff
@@ -82,66 +83,101 @@ private[filodb] final class IngestionActor(dataset: Dataset,
   override def postStop(): Unit = {
     super.postStop() // <- logs shutting down
     logger.info("Cancelling all streams and calling teardown")
-    streamSubscriptions.keys.foreach(stop(dataset.ref, _, ActorRef.noSender))
+    streamSubscriptions.keys.foreach(stopIngestion(_))
   }
 
-  /** All [[ShardCommand]] tasks are only started if the dataset
-    * and shard are valid for this ingester.
-    */
   def receive: Receive = LoggingReceive {
-    case e: StartShardIngestion        => start(e, sender())
-    case e: IngestRows                 => ingest(e)
-    case GetStatus                     => status(sender())
-    case StopShardIngestion(ds, shard) => stop(ds, shard, sender())
+    case GetStatus               => status(sender())
+    case e: IngestRows           => ingest(e)
+    case e: ShardIngestionState  => resync(e, sender())
   }
 
-  /** Guards that only this dataset's commands are acted upon.
-    * Handles initial memstore setup of dataset to shard.
-    * Also handles recovery process.
+  /**
+    * Compares the given shard mapper snapshot to the current set of shards being ingested and
+    * reconciles any differences. It does so by stopping ingestion for shards that aren't mapped
+    * to this node, and it starts ingestion for those that are.
     */
-  private def start(e: StartShardIngestion, origin: ActorRef): Unit =
-    if (invalid(e.ref)) handleInvalid(e, Some(origin)) else {
-      try memStore.setup(dataset, e.shard, storeConfig, downsample) catch {
-        case ShardAlreadySetup(ds, shard) =>
-          logger.warn(s"dataset=$ds shard=$shard already setup, skipping....")
-          return
-      }
+  private def resync(state: ShardIngestionState, origin: ActorRef): Unit = {
+    if (invalid(state.ref)) {
+      logger.error(s"$state is invalid for this ingester '${dataset.ref}'.")
+      return
+    }
 
-      val ingestion = for {
-        _ <- memStore.recoverIndex(dataset.ref, e.shard)
-        checkpoints <- memStore.metastore.readCheckpoints(dataset.ref, e.shard) }
-      yield {
-        if (checkpoints.isEmpty) {
-          // Start normal ingestion with no recovery checkpoint and flush group 0 first
-          normalIngestion(e.shard, None, 0, storeConfig.diskTTLSeconds)
+    if (state.version != 0 && state.version <= shardStateVersion) {
+      logger.info(s"Ignoring old ShardIngestionState version: ${state.version} <= $shardStateVersion")
+      return
+    }
+
+    // Start with the full set of all shards being ingested, and remove shards from this set
+    // which must continue being ingested.
+    val shardsToStop = HashSet() ++ streams.keySet
```

Comment:
It's probably adequate to just do `streams.keySet` as keySet is supposed to return some immutable Set.

Reply:
The goal here is to have a mutable copy of the set.